### ISSUE TITLE
SendGrid headers

### ISSUE THREAD

**Issue by bjunc** [thoughtbot/bamboo#445]
It doesn't appear that the SendGrid adapter supports headers.  Is that correct?  I see one-off support for a `reply-to`, but nothing anything else (eg. `X-Foo`).

Just to test my suspicion, I copied/modified the SendGridAdapter to merge `%Email{headers: headers}` into the body; which worked.  Happy to submit a PR, if headers are in-fact unsupported at the moment.

**Comment by maymillerricci** [comment]
@bjunc Thanks for posting! Which header(s) are you looking to pass through? Or were you thinking of making it possible to pass anything through, that map to the `headers` parameter for the API call? I think `reply-to` has one-off handling since it maps to a specific `reply_to` param. In any case, please do open a PR if there's something you are wanting for this that we don't yet have!

**Comment by michaelm244** [comment]
@maymillerricci I have this same problem - the headers that I'm interested in passing through are `In-Reply-To` and `References`. This allows you to reply to an email directly, assuming you have the message-ID of the email you'd like to reply to.

I think it's better to allow the user of the library to define what headers they want to send, rather than build one-off support for each specific header.

I just opened up a PR that adds this functionality, would love to hear any feedback/thoughts: https://github.com/thoughtbot/bamboo/pull/479

**Comment by bjunc** [comment]
Our use-case is similar. We're adding `In-Reply-To` as well as `References` to allow for message threading. We implemented something similar to the PR. Although, we simply put `put_headers` first in the pipeline so we don't worry about the `put_reply_to` (letting the user decide if they want to override the headers previously set).  For example:

```elixir
defp to_sendgrid_body(%Email{} = email, config) do
    %{}
    |> put_headers(email)
    |> put_from(email)
    |> put_personalization(email)
    |> put_reply_to(email)
    |> put_subject(email)
    |> put_content(email)
    |> put_template_id(email)
    |> put_attachments(email)
    |> put_categories(email)
    |> put_settings(config)
  end

defp put_headers(body, %Email{headers: headers})  when is_map(headers), 
  do: Map.put(body, :headers, headers)
```

Since the SendGrid docs go out of their way to say not to send certain reserved headers, we might want to make sure those aren't included in the header map (per the PR comment).

**Comment by maymillerricci** [comment]
This was implemented in #479.
